File: er_queue_sequential.cpp
#include <array>
#include <chrono>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <queue>
#include <random>
#include <string>
#include <vector>

namespace {
constexpr int RANDOM_SEED = 42;
constexpr int MIN_PRIORITY = 1;
constexpr int MAX_PRIORITY = 5;

const std::array<std::string, 6> PRIORITY_LABELS = {
    "",
    "Resuscitation",
    "Emergent",
    "Urgent",
    "Less Urgent",
    "Non-Urgent",
};

struct Patient {
    int priority;
    double arrival_time;
    int patient_id;
    double wait_time;
};

struct PatientCompare {
    bool operator()(const Patient& left, const Patient& right) const {
        if (left.priority != right.priority) {
            return left.priority > right.priority;
        }
        return left.arrival_time > right.arrival_time;
    }
};

std::string format_with_commas(unsigned long long value) {
    std::string digits = std::to_string(value);
    for (int i = static_cast<int>(digits.size()) - 3; i > 0; i -= 3) {
        digits.insert(static_cast<size_t>(i), ",");
    }
    return digits;
}

std::vector<Patient> generate_patients(size_t count) {
    std::mt19937 rng(RANDOM_SEED);
    std::uniform_int_distribution<int> priority_dist(MIN_PRIORITY, MAX_PRIORITY);
    const double simulation_window = 8.0 * 3600.0;
    std::uniform_real_distribution<double> arrival_dist(0.0, simulation_window);

    std::vector<Patient> patients;
    patients.reserve(count);

    for (size_t i = 0; i < count; ++i) {
        patients.push_back(Patient{
            priority_dist(rng),
            arrival_dist(rng),
            static_cast<int>(i + 1),
            0.0,
        });
    }

    return patients;
}

std::priority_queue<Patient, std::vector<Patient>, PatientCompare> build_priority_queue(
    const std::vector<Patient>& patients) {
    std::priority_queue<Patient, std::vector<Patient>, PatientCompare> queue;
    for (const auto& patient : patients) {
        queue.push(patient);
    }
    return queue;
}

void treat_patient(const Patient& patient) {
    (void)patient;
    const int limit = 1000;
    int primes_found = 0;
    int candidate = 2;

    while (primes_found < limit) {
        bool is_prime = true;
        int max_divisor = static_cast<int>(std::sqrt(candidate));
        for (int divisor = 2; divisor <= max_divisor; ++divisor) {
            if (candidate % divisor == 0) {
                is_prime = false;
                break;
            }
        }
        if (is_prime) {
            ++primes_found;
        }
        ++candidate;
    }
}

void run_simulation(size_t patient_count) {
    std::cout << "\n";
    std::cout << std::string(50, '=') << "\n";
    std::cout << "  ER QUEUE SIMULATION (SEQUENTIAL)\n";
    std::cout << std::string(50, '=') << "\n";
    std::cout << "Total Patients Generated: " << format_with_commas(patient_count) << "\n";
    std::cout << "Random Seed             : " << RANDOM_SEED << "\n";
    std::cout << "Treatment Work          : First 1,000 primes per patient\n";
    std::cout << std::string(50, '-') << "\n";

    std::cout << "Generating patients..." << std::flush;
    auto patients = generate_patients(patient_count);
    std::cout << " Done.\n";

    std::cout << "Building priority queue..." << std::flush;
    auto queue = build_priority_queue(patients);
    std::cout << " Done.\n";

    std::cout << "\n";
    std::cout << "Simulation Started...\n\n";

    std::vector<std::vector<double>> wait_times_by_priority(MAX_PRIORITY + 1);
    size_t patients_treated = 0;

    auto simulation_start = std::chrono::steady_clock::now();
    auto current_clock = simulation_start;

    while (!queue.empty()) {
        auto patient = queue.top();
        queue.pop();

        auto normalised_arrival = simulation_start +
            std::chrono::duration<double>(patient.arrival_time);
        double wait_time = 0.0;
        if (current_clock > normalised_arrival) {
            wait_time = std::chrono::duration<double>(current_clock - normalised_arrival).count();
        }
        patient.wait_time = wait_time;
        wait_times_by_priority[patient.priority].push_back(wait_time);

        treat_patient(patient);
        ++patients_treated;

        current_clock = std::chrono::steady_clock::now();
    }

    auto simulation_end = std::chrono::steady_clock::now();
    double total_time =
        std::chrono::duration<double>(simulation_end - simulation_start).count();
    double throughput = total_time > 0.0
        ? static_cast<double>(patients_treated) / total_time
        : 0.0;

    std::cout << "Simulation Complete.\n\n";
    std::cout << "Total Execution Time: " << std::fixed << std::setprecision(2)
              << total_time << " seconds\n\n";
    std::cout << "Average Wait Time per Priority:\n";

    for (int priority = MIN_PRIORITY; priority <= MAX_PRIORITY; ++priority) {
        const auto& times = wait_times_by_priority[priority];
        if (!times.empty()) {
            double sum = 0.0;
            for (double value : times) {
                sum += value;
            }
            double average = sum / static_cast<double>(times.size());
            std::cout << "  - Priority " << priority << " (" << std::setw(14)
                      << PRIORITY_LABELS[priority] << "): " << std::fixed
                      << std::setprecision(4) << average << "s  ["
                      << format_with_commas(times.size()) << " patients]\n";
        } else {
            std::cout << "  - Priority " << priority << ": no patients\n";
        }
    }

    auto throughput_value = static_cast<unsigned long long>(std::llround(throughput));
    std::cout << "\nThroughput: " << format_with_commas(throughput_value)
              << " patients/second\n";
    std::cout << std::string(50, '=') << "\n\n";
}
}  // namespace

int main(int argc, char* argv[]) {
    size_t patient_count = 10000;

    if (argc > 1) {
        try {
            long long parsed = std::stoll(argv[1]);
            if (parsed <= 0) {
                throw std::invalid_argument("non-positive");
            }
            patient_count = static_cast<size_t>(parsed);
        } catch (const std::exception&) {
            std::cout << "[ERROR] Patient count must be a positive integer. Got: "
                      << argv[1] << "\n";
            return 1;
        }
    }

    run_simulation(patient_count);
    return 0;
}

File: er_queue_sequential.py
import heapq
import random
import time
import sys
import math
from dataclasses import dataclass, field
from typing import List, Tuple

RANDOM_SEED = 42          
PRIORITY_LABELS = {
    1: "Resuscitation",
    2: "Emergent",
    3: "Urgent",
    4: "Less Urgent",
    5: "Non-Urgent",
}


# ─────────────────────────────────────────────
#  PATIENT DATA CLASS
# ─────────────────────────────────────────────

@dataclass(order=True)
class Patient:
    """
    Represents a single ER patient.

    The dataclass is ordered so that heapq can compare patients.
    Sort key is (priority, arrival_time) — lower priority number = higher urgency,
    and among equal priorities, earlier arrival is seen first (FIFO).
    """
    priority: int                          # 1–5 (1 is most urgent)
    arrival_time: float                    # simulated timestamp in seconds
    patient_id: int = field(compare=False) # unique ID, excluded from ordering
    wait_time: float = field(default=0.0, compare=False)  # filled in after treatment


# ─────────────────────────────────────────────
#  PATIENT GENERATOR
# ─────────────────────────────────────────────

def generate_patients(n: int) -> List[Patient]:
    """
    Generate N patients with:
      - Unique sequential patient IDs (P0001, P0002, …)
      - Random triage priority (1–5)
      - Simulated arrival time (seconds from simulation start, spread over 8 hours)

    """
    random.seed(RANDOM_SEED)

    patients = []
    simulation_window = 8 * 3600  # 8-hour ER shift in seconds

    for i in range(n):
        priority = random.randint(1, 5)
        # Arrivals distributed across the shift; not guaranteed monotonic —
        # the queue will sort them correctly regardless.
        arrival_time = random.uniform(0, simulation_window)
        patients.append(Patient(
            priority=priority,
            arrival_time=arrival_time,
            patient_id=i + 1,
        ))

    return patients


# ─────────────────────────────────────────────
#  PRIORITY QUEUE BUILDER
# ─────────────────────────────────────────────

def build_priority_queue(patients: List[Patient]) -> list:
    """
    Push all patients into a min-heap.

    heapq in Python is a min-heap, and Patient ordering is (priority, arrival_time),
    so Priority-1 patients with the earliest arrival bubble to the top.
    This satisfies the spec:
      • Lower number = higher urgency = served first
      • FIFO among same-priority patients (earlier arrival_time wins)
    """
    heap = []
    for patient in patients:
        heapq.heappush(heap, patient)
    return heap


# ─────────────────────────────────────────────
#  TREATMENT COMPUTATION
#  (CPU-bound math loop)
# ─────────────────────────────────────────────

def treat_patient(patient: Patient) -> None:
    """
    Simulate treatment with a CPU-bound mathematical operation.

    We calculate the first 1,000 prime numbers using a trial-division
    sieve. This is the EXACT same algorithm that Noel must implement in C++
    so that both simulations spend the same amount of CPU work per patient,
    making execution-time comparisons fair.

    DO NOT use time.sleep() here — sleep yields the CPU and does not
    reflect actual computation performance.
    """
    limit = 1_000
    primes_found = 0
    candidate = 2

    while primes_found < limit:
        is_prime = True
        for divisor in range(2, int(math.isqrt(candidate)) + 1):
            if candidate % divisor == 0:
                is_prime = False
                break
        if is_prime:
            primes_found += 1
        candidate += 1


# ─────────────────────────────────────────────
#  SEQUENTIAL SIMULATION RUNNER
# ─────────────────────────────────────────────

def run_simulation(n_patients: int) -> None:
    """
    Main simulation loop.

    A single 'doctor' (the main thread) pulls patients from the priority
    queue one at a time and treats them sequentially. This is the Phase 1
    baseline that Phase 2 (Edricka's parallel version) will improve upon.
    """

    print()
    print("=" * 50)
    print("  ER QUEUE SIMULATION (SEQUENTIAL)")
    print("=" * 50)
    print(f"Total Patients Generated: {n_patients:,}")
    print(f"Random Seed             : {RANDOM_SEED}")
    print(f"Treatment Work          : First 1,000 primes per patient")
    print("-" * 50)

    # ── Step 1: Generate patients ──────────────────
    print("Generating patients...", end=" ", flush=True)
    patients = generate_patients(n_patients)
    print("Done.")

    # ── Step 2: Build priority queue ───────────────
    print("Building priority queue...", end=" ", flush=True)
    queue = build_priority_queue(patients)
    print("Done.")

    # ── Step 3: Run sequential treatment ───────────
    print()
    print("Simulation Started...")
    print()

    # Metric accumulators
    wait_times_by_priority: dict[int, List[float]] = {p: [] for p in range(1, 6)}
    patients_treated = 0

    simulation_start = time.perf_counter()
    current_clock = simulation_start  # tracks when the doctor becomes free

    while queue:
        patient = heapq.heappop(queue)

        # The doctor becomes available at current_clock.
        # The patient's wait = max(0, current_clock - their simulated arrival).
        # (We normalise arrival_time relative to simulation_start for metric purposes.)
        normalised_arrival = simulation_start + patient.arrival_time
        wait = max(0.0, current_clock - normalised_arrival)
        patient.wait_time = wait
        wait_times_by_priority[patient.priority].append(wait)

        # Treat the patient (CPU-bound work)
        treat_patient(patient)
        patients_treated += 1

        # Doctor is now free
        current_clock = time.perf_counter()

    simulation_end = time.perf_counter()

    # ── Step 4: Compute & print metrics ───────────
    total_time = simulation_end - simulation_start
    throughput = patients_treated / total_time if total_time > 0 else 0

    print("Simulation Complete.")
    print()
    print(f"Total Execution Time: {total_time:.2f} seconds")
    print()
    print("Average Wait Time per Priority:")
    for priority in range(1, 6):
        times = wait_times_by_priority[priority]
        if times:
            avg = sum(times) / len(times)
            label = PRIORITY_LABELS[priority]
            print(f"  - Priority {priority} ({label:>14}): {avg:.4f}s  [{len(times):,} patients]")
        else:
            print(f"  - Priority {priority}: no patients")

    print()
    print(f"Throughput: {throughput:,.0f} patients/second")
    print("=" * 50)
    print()


# ─────────────────────────────────────────────
#  ENTRY POINT
# ─────────────────────────────────────────────

if __name__ == "__main__":
    # Accept optional CLI argument for patient count
    if len(sys.argv) > 1:
        try:
            n = int(sys.argv[1])
            if n <= 0:
                raise ValueError
        except ValueError:
            print(f"[ERROR] Patient count must be a positive integer. Got: {sys.argv[1]}")
            sys.exit(1)
    else:
        n = 10_000  # sensible default for a quick test run

    run_simulation(n)
